core/lib/state/src/cache/sequential_cache.rs

```diff
@@ -0,0 +1,150 @@
+use std::collections::VecDeque;
+
+use crate::cache::metrics::{Method, RequestOutcome, METRICS};
+
+/// A generic cache structure for storing key-value pairs in sequential order.
+/// It allows for non-unique keys and supports efficient retrieval of values based on a key
+/// threshold. The cache maintains a specified maximum capacity, removing the oldest entries
+/// as new ones are added.
+///
+/// Usage example: storing mempool transactions and querying them by the received_at field
+/// (clients maintain a cursor based on this key and poll the data structure for newer transactions)
+///
+///
+/// K: Type of the ordered, potentially non-unique key. Example: Transaction's received_at field.
+/// V: Type of the value associated with each key. Example: Transaction's hash.
+#[derive(Debug, Clone)]
+pub struct SequentialCache<K, V> {
+    name: &'static str,
+    data: VecDeque<(K, V)>,
+    capacity: usize,
+}
+
+impl<K: Ord + Copy, V: Clone> SequentialCache<K, V> {
```

Comment:
Nit: `Copy` requirement seems overly strict here, is it intentional or `Clone` would suffice? (Though I think even `Clone` can be avoided on both `K` and `V` if we change the signatures to work with references, but it'll alter the UX so dunno if that aligns with your point of view).